skopt/searchcv.py

```diff
@@ -0,0 +1,448 @@
+            names (strings) and values are skopt.space.Dimension instances,
+            one of Real, Integer or Categorical.
+
+        Returns
+        -------
+        optimizer: Instance of the `Optimizer` class used for for search
+            in some parameter space.
+
+        """
+        # convert search space from dict to list
+        dimensions = [params_space[k] for k in sorted(params_space.keys())]
+
+        kwargs = self.optimizer_kwargs.copy()
+        kwargs['dimensions'] = dimensions
+        optimizer = Optimizer(**kwargs)
+
+        return optimizer
+
+    def _skopt_to_dict(self, params_space, params):
+        """Converts list of parameter values into the dictionary with
+        keys as parameter names and corresponding values of parameter.
+
+        Parameters
+        ----------
+        params_space : dict
+            Represents parameter search space. The keys are parameter
+            names (strings) and values are skopt.space.Dimension instances,
+            one of Real, Integer or Categorical.
+
+        params : list
+            Parameter values as list. The order of parameters in the list
+            is given by sorted(params_space.keys()).
+
+        Returns
+        -------
+        params_dict: dictionary with parameter values.
+        """
+        params_dict = {
+            k: v for k, v in zip(sorted(params_space.keys()), params)
+        }
+        return params_dict
+
+    def step(self, X, y, param_space, groups=None, n_jobs=1):
+        """Generates n_jobs parameters and evaluates corresponding
+        estimators in parallel.
+
+        Having a separate function for a single step for search allows to
+        save easily checkpoints for the parameter search and restore from
+        possible failures.
+
+        Parameters
+        ----------
+        X : array-like or sparse matrix, shape = [n_samples, n_features]
+            The training input samples. Internally, it will be converted to
+            ``dtype=np.float32`` and if a sparse matrix is provided
+            to a sparse ``csc_matrix``.
+
+        y : array-like, shape = [n_samples] or [n_samples, n_outputs]
+            The target values (class labels) as integers or strings.
+
+        params_space : dict
+            Represents parameter search space. The keys are parameter
+            names (strings) and values are skopt.space.Dimension instances,
+            one of Real, Integer or Categorical.
+
+        params : list
+            Parameter values as list. The order of parameters in the list
+            is given by sorted(params_space.keys()).
+
+        Returns
+        -------
+        params_dict: dictionary with parameter values.
+        """
+
+        # convert n_jobst to int > 0 if necessary
+        if n_jobs < 0:
+            n_jobs = max(1, cpu_count() + n_jobs + 1)
+
+        # use the cached optimizer for particular parameter space
+        key = str(param_space)
```

Comment:
Each optimizer is associated to one of the spaces defined by the user (as a list). So could we make `self.optimizer_` a list as well? Where the order of optimizers corresponds to the order of spaces? Only thought of this yesterday night while pondering how likely it is that we will get subtle bugs because of using `str()` (In my experience constructing a way to hash an unhashable object is the path to pain and subtle bugs)